### ISSUE TITLE
Invalid PartialEq and Hash implementation for unions

### ISSUE THREAD

**Issue by asomers** [rust-lang/rust#97606]
`union semun`, in unix/bsd/apple/mod.rs, has three fields.  On LP64, these fields will have unequal sizes.  But its `PartialEq` and `Hash` implementations only check the smallest field.  So the following code should fail (untested, because I don't have access to a Mac):

```rust
use libc::semun;

#[test]
fn semun_eq() {
    let x = semun{ val: 0x44556677 };
    let y = semun{ array: 0x0011223344556677 };
    assert!(x != y);
}
```

A similar situation holds on Linux for `__c_anonymous_ptrace_syscall_info_data`.  That union has three members of different sizes, and its `PartialEq` implementation will return turn as long as *any* pair of fields compare equal.  Effectively, that means it's only comparing the smallest field.

There are other unions in libc, but I haven't audited them all.

**Comment by RalfJung** [comment]
Shouldn't this be reported against https://github.com/rust-lang/libc ? Sounds like a bug in libc to me, specifically [around here](https://github.com/rust-lang/libc/blob/99756b1a974d293a3c61d13f5cf4affe6d0998fc/src/unix/bsd/apple/mod.rs#L1304).

OTOH, that code cannot just compare `buf` or `array`, since the bytes outside of `val` might be uninitialized. I don't think it is possible to have a reasonable implementation of `PartialEq` for this type.

**Comment by asomers** [comment]
Oh shoot, I meant to file it to that repository.  Sorry for the noise.
